File: mkdocs/snippets/devbook/accounts/query_balance.py
import json
import os
import urllib.request

NODE_URL = os.getenv('NODE_URL', 'https://reference.symboltest.net:3001')
print(f'Using node {NODE_URL}')


def get_account_info(account_identifier):  # [>step-1]
	"""
	Fetch account information by address or public key.

	Args:
		account_identifier: The account address or public key

	Returns:
		Dictionary containing the account information
	"""
	account_path = f'/accounts/{account_identifier}'
	try:
		account_url = f'{NODE_URL}{account_path}'
		with urllib.request.urlopen(account_url) as response:
			account_info = json.loads(response.read().decode())
			return account_info['account']
	except urllib.error.HTTPError as err:
		if err.status == 404:
			print(f'Address does not exist: {err}')
		elif err.status == 409:
			print(f'Address is not properly formatted: {err}')
		else:
			print(f'Unexpected error: {err}')
		raise SystemExit(1) from err  # [<step-1]


def get_mosaic_names(mosaic_ids):  # [>step-2]
	"""
	Fetch friendly names for a set of mosaics.

	Args:
		mosaic_ids: List of mosaic IDs as integers

	Returns:
		Dictionary mapping mosaic IDs to their namespace names
	"""
	mosaic_ids_hex = [f'{mosaic_id:016X}' for mosaic_id in mosaic_ids]
	request_body = json.dumps({'mosaicIds': mosaic_ids_hex}).encode()
	request = urllib.request.Request(
		f'{NODE_URL}/namespaces/mosaic/names',
		data=request_body,
		headers={'Content-Type': 'application/json'}
	)
	with urllib.request.urlopen(request) as response:
		names_info = json.loads(response.read().decode())
		# Build a dictionary mapping mosaic IDs to their names
		names_map = {}
		for entry in names_info['mosaicNames']:
			entry_id = int(entry['mosaicId'], 16)
			names_map[entry_id] = entry['names']
		return names_map  # [<step-2]


def get_mosaics_info(mosaic_ids):  # [>step-3]
	"""
	Fetch information for multiple mosaics in a single request.

	Args:
		mosaic_ids: List of mosaic IDs as integers

	Returns:
		Dictionary mapping mosaic IDs to their properties
	"""
	mosaic_ids_hex = [f'{mosaic_id:016X}' for mosaic_id in mosaic_ids]
	request_body = json.dumps({'mosaicIds': mosaic_ids_hex}).encode()
	request = urllib.request.Request(
		f'{NODE_URL}/mosaics',
		data=request_body,
		headers={'Content-Type': 'application/json'}
	)
	with urllib.request.urlopen(request) as response:
		mosaics_info = json.loads(response.read().decode())
		# Build a dictionary mapping mosaic IDs to their properties
		mosaics_map = {}
		for entry in mosaics_info:
			entry_id = int(entry['mosaic']['id'], 16)
			mosaics_map[entry_id] = entry['mosaic']
		return mosaics_map  # [<step-3]


def format_amount(amount, divisibility):  # [>step-4]
	"""
	Format an atomic amount with decimal places.

	Args:
		amount: The atomic amount as an integer
		divisibility: Number of decimal places

	Returns:
		Formatted amount as a string
	"""
	if divisibility == 0:
		return str(amount)
	whole_part = amount // (10 ** divisibility)
	fractional_part = amount % (10 ** divisibility)
	return f'{whole_part}.{fractional_part:0{divisibility}d}'  # [<step-4]


# The account address to query [>step-5]
ADDRESS = os.getenv(
	'ADDRESS', 'TBIL6D6RURP45YQRWV6Q7YVWIIPLQGLZQFHWFEQ')
print(f'Fetching account information from {ADDRESS}')

try:
	# Get account information
	account = get_account_info(ADDRESS)

	# Display balances for all mosaics the account holds
	account_mosaics = account['mosaics']
	if not account_mosaics:
		print('Account holds no mosaics')
	else:
		print(f'Account holds {len(account_mosaics)} mosaic(s):')

		# Fetch mosaic properties and names for all mosaics
		acc_mosaic_ids = [int(m['id'], 16) for m in account_mosaics]
		acc_mosaic_names = get_mosaic_names(acc_mosaic_ids)
		acc_mosaics_info = get_mosaics_info(acc_mosaic_ids)

		for mosaic_entry in account_mosaics:
			mosaic_id = int(mosaic_entry['id'], 16)
			balance = int(mosaic_entry['amount'])

			# Get mosaic properties
			info = acc_mosaics_info[mosaic_id]
			mosaic_divisibility = info['divisibility']

			# Format and display the balance
			formatted_balance = format_amount(balance, mosaic_divisibility)
			mosaic_id_hex = f'0x{mosaic_id:016X}'

			# Display mosaic ID and names (if available)
			names = acc_mosaic_names.get(mosaic_id, [])
			if names:
				names_str = ', '.join(names)
				print(f'- Mosaic {mosaic_id_hex} ({names_str})')
			else:
				print(f'- Mosaic {mosaic_id_hex}')

			print(f'  Balance: {formatted_balance}')
			print(f'  Balance (atomic): {balance}')
			print(f'  Divisibility: {mosaic_divisibility}')
except urllib.error.URLError as e:
	print(e.reason)  # [<step-5]

File: mkdocs/snippets/devbook/accounts/query_balance.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.1

import java.io.IOException;
import java.math.BigInteger;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

final class QueryBalance {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final String NODE_URL = System.getenv().getOrDefault(
		"NODE_URL", "https://reference.symboltest.net:3001");

	// [>step-1]
	/**
	 * Fetch account information by address or public key.
	 */
	private static JsonNode getAccountInfo(
		final String accountIdentifier
	) throws IOException, InterruptedException {
		final String accountPath = String.format(
			"/accounts/%s", accountIdentifier);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(NODE_URL + accountPath)).GET().build();
		final HttpResponse<String> response = HttpClient.newHttpClient()
			.send(request, BodyHandlers.ofString());

		if (200 != response.statusCode()) {
			if (404 == response.statusCode())
				System.out.printf(
					"Address does not exist: %s%n", response.body());
			else if (409 == response.statusCode())
				System.out.printf(
					"Address is not properly formatted: %s%n",
					response.body());
			else
				System.out.printf("Unexpected error: %s%n",
					response.body());
			System.exit(1);
		}

		return JSON_MAPPER.readTree(response.body()).get("account");
	} // [<step-1]

	// [>step-2]
	/**
	 * Fetch friendly names for a set of mosaics.
	 */
	private static Map<BigInteger, List<String>> getMosaicNames(
		final List<BigInteger> mosaicIds
	) throws IOException, InterruptedException {
		final ObjectNode requestBody = JSON_MAPPER.createObjectNode();
		final ArrayNode mosaicIdsHex = requestBody.putArray("mosaicIds");
		for (final BigInteger mosaicId : mosaicIds)
			mosaicIdsHex.add(String.format("%016X", mosaicId));
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(NODE_URL + "/namespaces/mosaic/names"))
			.header("Content-Type", "application/json")
			.POST(HttpRequest.BodyPublishers.ofString(
				requestBody.toString()))
			.build();
		final HttpResponse<String> response = HttpClient.newHttpClient()
			.send(request, BodyHandlers.ofString());
		final JsonNode namesInfo = JSON_MAPPER.readTree(response.body());

		// Build a map from mosaic IDs to their names
		final Map<BigInteger, List<String>> namesMap = new HashMap<>();
		for (final JsonNode entry : namesInfo.get("mosaicNames")) {
			final BigInteger mosaicId = new BigInteger(
				entry.get("mosaicId").asText(), 16);
			final List<String> names = new ArrayList<>();
			for (final JsonNode name : entry.get("names"))
				names.add(name.asText());
			namesMap.put(mosaicId, names);
		}
		return namesMap;
	} // [<step-2]

	// [>step-3]
	/**
	 * Fetch information for multiple mosaics in a single request.
	 */
	private static Map<BigInteger, JsonNode> getMosaicsInfo(
		final List<BigInteger> mosaicIds
	) throws IOException, InterruptedException {
		final ObjectNode requestBody = JSON_MAPPER.createObjectNode();
		final ArrayNode mosaicIdsHex = requestBody.putArray("mosaicIds");
		for (final BigInteger mosaicId : mosaicIds)
			mosaicIdsHex.add(String.format("%016X", mosaicId));
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(NODE_URL + "/mosaics"))
			.header("Content-Type", "application/json")
			.POST(HttpRequest.BodyPublishers.ofString(
				requestBody.toString()))
			.build();
		final HttpResponse<String> response = HttpClient.newHttpClient()
			.send(request, BodyHandlers.ofString());
		final JsonNode mosaicsInfo = JSON_MAPPER.readTree(
			response.body());

		// Build a map from mosaic IDs to their properties
		final Map<BigInteger, JsonNode> mosaicsMap = new HashMap<>();
		for (final JsonNode entry : mosaicsInfo) {
			final JsonNode mosaic = entry.get("mosaic");
			final BigInteger mosaicId = new BigInteger(
				mosaic.get("id").asText(), 16);
			mosaicsMap.put(mosaicId, mosaic);
		}
		return mosaicsMap;
	} // [<step-3]

	// [>step-4]
	/**
	 * Format an atomic amount with decimal places.
	 */
	private static String formatAmount(
		final BigInteger amount,
		final int divisibility
	) {
		if (0 == divisibility)
			return amount.toString();

		final BigInteger divisor = BigInteger.TEN.pow(divisibility);
		final BigInteger[] parts = amount.divideAndRemainder(divisor);
		return String.format(
			"%s.%0" + divisibility + "d", parts[0], parts[1]);
	} // [<step-4]

	public static void main(final String[] args) {
		new QueryBalance().run();
	}

	private void run() {
		System.out.printf("Using node %s%n", NODE_URL);

		// The account address to query [>step-5]
		final String address = System.getenv().getOrDefault(
			"ADDRESS", "TBIL6D6RURP45YQRWV6Q7YVWIIPLQGLZQFHWFEQ");
		System.out.printf("Fetching account information from %s%n",
			address);

		try {
			// Get account information
			final JsonNode account = getAccountInfo(address);

			// Display balances for all mosaics the account holds
			final JsonNode accountMosaics = account.get("mosaics");
			if (accountMosaics.isEmpty()) {
				System.out.println("Account holds no mosaics");
			} else {
				System.out.printf("Account holds %d mosaic(s):%n",
					accountMosaics.size());

				// Fetch mosaic properties and names for all mosaics
				final List<BigInteger> mosaicIds = new ArrayList<>();
				for (final JsonNode mosaicEntry : accountMosaics)
					mosaicIds.add(new BigInteger(
						mosaicEntry.get("id").asText(), 16));
				final Map<BigInteger, List<String>> mosaicNames =
					getMosaicNames(mosaicIds);
				final Map<BigInteger, JsonNode> mosaicsInfo =
					getMosaicsInfo(mosaicIds);

				for (final JsonNode mosaicEntry : accountMosaics) {
					final BigInteger mosaicId = new BigInteger(
						mosaicEntry.get("id").asText(), 16);
					final BigInteger balance = new BigInteger(
						mosaicEntry.get("amount").asText());

					// Get mosaic properties
					final JsonNode info = mosaicsInfo.get(mosaicId);
					final int divisibility = info.get("divisibility")
						.asInt();

					// Format and display the balance
					final String formattedBalance = formatAmount(
						balance, divisibility);
					final String mosaicIdHex = String.format(
						"0x%016X", mosaicId);

					// Display mosaic ID and names (if available)
					final List<String> names = mosaicNames.getOrDefault(
						mosaicId, List.of());
					if (names.isEmpty())
						System.out.printf("- Mosaic %s%n", mosaicIdHex);
					else
						System.out.printf("- Mosaic %s (%s)%n",
							mosaicIdHex,
							String.join(", ", names));

					System.out.printf("  Balance: %s%n",
						formattedBalance);
					System.out.printf("  Balance (atomic): %s%n",
						balance);
					System.out.printf("  Divisibility: %d%n",
						divisibility);
				}
			}
		} catch (final Exception ex) {
			System.out.println(ex.getMessage());
		} // [<step-5]
	}
}
